instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressionStrategy.java

```diff
@@ -0,0 +1,124 @@
+      Arrays.asList(SpanKey.CONSUMER));
+
+  static SpanSuppressionStrategy from(List<SpanKey> clientSpanKeys) {
+    if (clientSpanKeys.isEmpty()) {
+      return new CompositeStrategy(
+          NeverSuppress.INSTANCE,
+          SERVER_STRATEGY,
+          CONSUMER_STRATEGY);
+    }
+    return new CompositeStrategy(
+        new SuppressIfSameType(clientSpanKeys),
+        SERVER_STRATEGY,
+        CONSUMER_STRATEGY);
+  }
+
+  abstract Context storeInContext(SpanKind spanKind, Context context, Span span);
+
+  abstract boolean shouldSuppress(SpanKind spanKind, Context parentContext);
+
+  static final class SuppressIfSameType extends SpanSuppressionStrategy {
+
+    private final List<SpanKey> outgoingSpanKeys;
+
+    SuppressIfSameType(List<SpanKey> outgoingSpanKeys) {
+      this.outgoingSpanKeys = outgoingSpanKeys;
+    }
+
+    @Override
+    Context storeInContext(SpanKind spanKind, Context context, Span span) {
+      for (SpanKey outgoingSpanKey : outgoingSpanKeys) {
+        context = outgoingSpanKey.with(context, span);
+      }
+      return context;
+    }
+
+    @Override
+    boolean shouldSuppress(SpanKind spanKind, Context parentContext) {
+        for (SpanKey outgoingSpanKey : outgoingSpanKeys) {
+          if (outgoingSpanKey.fromContextOrNull(parentContext) == null) {
+            return false;
+          }
+        }
+        return true;
+    }
+  }
+
+  static final class NeverSuppress extends SpanSuppressionStrategy {
+
+    private static final SpanSuppressionStrategy INSTANCE = new NeverSuppress();
+
+    @Override
+    Context storeInContext(SpanKind spanKind, Context context, Span span) {
+      return context;
+    }
+
+    @Override
+    boolean shouldSuppress(SpanKind spanKind, Context parentContext) {
+      return false;
+    }
+  }
+
+  static final class CompositeStrategy extends SpanSuppressionStrategy {
+    private final SpanSuppressionStrategy clientStrategy;
+    private final SpanSuppressionStrategy serverStrategy;
+    private final SpanSuppressionStrategy consumerStrategy;
+
+    public CompositeStrategy(SpanSuppressionStrategy client, SpanSuppressionStrategy server,
+        SpanSuppressionStrategy consumer) {
+      this.clientStrategy = client;
+      this.serverStrategy = server;
+      this.consumerStrategy = consumer;
+    }
+
+    @Override
+    Context storeInContext(SpanKind spanKind, Context context, Span span) {
+      switch (spanKind) {
+        case CLIENT:
+        case PRODUCER:
+          return clientStrategy.storeInContext(spanKind, context, span);
+        case SERVER:
+          return serverStrategy.storeInContext(spanKind, context, span);
+        case CONSUMER:
+          return consumerStrategy.storeInContext(spanKind, context, span);
+        case INTERNAL:
+          return context;
+      }
+      return context;
+    }
+
+    @Override
+    boolean shouldSuppress(SpanKind spanKind, Context parentContext) {
+      switch (spanKind) {
```

Comment:
I'd love to remove the association between span kind and span suppression at some point in the future - but I guess that's still impossible at this point. Maybe when we split our `AttributesExtractor` implementations into server/client? We've been thinking of doing that for HTTP, we could as well do the same thing for other conventions.

Reply:
it seems we'd have to specialize instrumenters (so specific instumenter instance work for one span kind) to remove the association between kind/suppression. Or perhaps, if we move `CompositeStrategy` into the instrumenter and it would have 4 of them - one per kind, it would not remove the association but maybe a bit cleaner.
Either way, a combination of type + kind is important for suppression (i.e. server HTTP span should not suppress client HTTP span) 

Another thought is that all strategies can be the same regardless of kind - suppress same type+kind (or suppress the same kind if type disabled) removes the association. I assume it is rare to have nested SERVERs of different types and we can easily allow that.